src/render/draw_collision_debug.js

```diff
@@ -41,9 +50,165 @@ function drawCollisionDebugGeometry(painter: Painter, sourceCache: SourceCache,
             buffers.segments, null, painter.transform.zoom, null, null,
             buffers.collisionVertexBuffer);
     }
+
+    if (isText)
+        drawCollisionCircles(painter, sourceCache, layer, coords, translate, translateAnchor);
 }
 
-function drawCollisionDebug(painter: Painter, sourceCache: SourceCache, layer: StyleLayer, coords: Array<OverscaledTileID>, translate: [number, number], translateAnchor: 'map' | 'viewport', isText: boolean) {
-    drawCollisionDebugGeometry(painter, sourceCache, layer, coords, false, translate, translateAnchor, isText);
-    drawCollisionDebugGeometry(painter, sourceCache, layer, coords, true, translate, translateAnchor, isText);
+function drawCollisionCircles(painter: Painter, sourceCache: SourceCache, layer: StyleLayer, coords: Array<OverscaledTileID>, translate: [number, number], translateAnchor: 'map' | 'viewport') {
+    // Collision circle rendering is done by using simple shader batching scheme where dynamic properties of
+    // circles are passed to the GPU using shader uniforms. Circles are first encoded into 4-component vectors
+    // (center_x, center_y, radius, flag) and then uploaded in batches as "uniform vec4 u_quads[N]".
+    // Vertex data is just a collection of incremental index values pointing to the quads-array.
+    //
+    // If one quad uses 4 vertices then all required values can be deduced from the index value:
+    //   int quad_idx = int(vertex.idx / 4);
+    //   int corner_idx = int(vertex.idx % 4);
+    //
+    // OpenGL ES 2.0 spec defines that the maximum number of supported vertex uniform vectors (vec4) should be
+    // at least 128. Choosing a safe value 64 for the quad array should leave enough space for rest of the
+    // uniform variables.
+    const maxQuadsPerDrawCall = 64;
```

Comment:
Should we inject a define in the circleProgram prelude for that value https://github.com/mapbox/mapbox-gl-js/blob/master/src/render/program.js#L40? 

That would prevent mismatch with the variable defined in src/shaders/collision_circle.vertex.glsl.

Reply:
That's a great idea and something we should definitely do. `ProgramConfiguration` doesn't seem to be flexible enough for this purpose though as it's built around binding style and tile layer properties into uniforms and vertex attributes. Do you have some other approach in mind?

Reply:
Thanks for checking, if that requires some extra work it's fine to leave as is for the time being. We can revisit and extend `ProgramConfiguration` for that type of generic shader config injection when it shows problems.